src/charm.py

```diff
@@ -397,20 +396,22 @@ def _get_unit_id(self):
     def _configure_cni(self):
         """Configure calico cni."""
         self.unit.status = MaintenanceStatus("Configuring Calico CNI.")
-        ip_versions = self._get_ip_versions()
-        ip6 = "autodetect" if 6 in ip_versions else "none"
-        self.cni_options.update(
-            {
-                "kubeconfig_path": "/opt/calicoctl/kubeconfig",
-                "mtu": self._get_mtu(),
-                "assign_ipv4": "true" if 4 in ip_versions else "false",
-                "assign_ipv6": "true" if 6 in ip_versions else "false",
-                "IP6": ip6,
-            }
-        )
+        self.calico_manifests.cni_config = self._get_cni_config()
```

Comment:
i feel like i'm missing is the subtle context that makes this different. 

Charm starts up and that `cni_config` in `CalicoManifest` is `{}`

until the `_configure_cni(self)` comes along and either updates or sets it for the first time.  

what's really changing?

Reply:
I'm also a puzzled by this. It looks to me like in the original code, `self.cni_config` and `self.calico_manifests.cni_config` are both references to the same dictionary. In `_configure_cni`, when the charm calls `self.cni_config.update`, those updates should appear in `self.calico_manifests.cni_config` too.

This PR makes the mutation happen at a higher level, replacing the whole dictionary instead of updating an existing one. But shouldn't the end result be the same?

Reply:
Actually, what stands out to me is that we're relying on `_configure_cni` to set `cni_config` prior to applying the manifests. But there are two code paths ([here](https://github.com/charmed-kubernetes/charm-calico/blob/b21b59e305015b4ae42a1dd8cd42fdc2a96178e5/src/charm.py#L91) and [here](https://github.com/charmed-kubernetes/charm-calico/blob/b21b59e305015b4ae42a1dd8cd42fdc2a96178e5/src/charm.py#L101)) that call `apply_manifests` without having previously called `_configure_cni`. Could the nulls be coming from those two cases?

Reply:
Yes, you're right @Cynerva. I think we should reconfigure the CNI in those code paths. I've noticed while testing the DualStack functionality that sometimes the etcd cluster refreshes and triggers the relation change. Therefore, addressing both reconfigurations should be the best option here. Thank you @addyess and @Cynerva for your reviews. I have pushed a new version that addresses this issue and also rolls back the previous change.